main/src/library/EncodingWriter.flix

```diff
@@ -0,0 +1,115 @@
+
+instance Writable[EncodingWriter[t]] with Writable[t] where Writable.Elm[t] ~ Int8 {
+
+    type Elm = Char
+
+    type Aef = Writable.Aef[t] + IO
+
+    pub def write(buffer: Array[Char, r], writer: EncodingWriter[t]): Result[IoError, Int32] \ r + Writable.Aef[t] + IO = EncodingWriter.write(buffer, writer)
+
+}
+
+mod EncodingWriter {
+    import java.nio.charset.CharsetEncoder
+    import java.nio.charset.CodingErrorAction
+    import java.nio.ByteBuffer
+    import java.nio.CharBuffer
+
+    ///
+    /// Wraps `byteWriter` in an `EncodingWriter`.
+    ///
+    pub def wrap(charset: CharacterSet, byteWriter: t): EncodingWriter[t] \ IO =
+        let CharacterSet.CharacterSet(cs) = charset;
+        EncodingWriter({
+            encoder = cs.newEncoder().onMalformedInput(CodingErrorAction.REPLACE).onUnmappableCharacter(CodingErrorAction.REPLACE),
+            byteWriter = byteWriter
+        })
+
+    ///
+    /// Reads `k` items from `b` and writes them into the underlying resource.
+    ///
+    /// Returns `Ok(k)` to signify that `k` items were successfully read and written to `r`.
+    ///
+    /// Guarantees that `0 <= k <= length(b)`.
+    ///
+    /// Returns `Err(e)` if some underlying I/O error occurs.
+    ///
+    pub def write(src: Array[Char, r], writer: EncodingWriter[t]): Result[IoError, Int32] \ IO + r + Writable.Aef[t] with Writable[t] where Writable.Elm[t] ~ Int8 =
+        let EncodingWriter.EncodingWriter({ encoder, byteWriter }) = writer;
+        region rc {
+            let dest = encoder |> encode(rc, src);
+            forM(_ <- byteWriter |> writeBytes(rc, dest)) yield src |> Array.length
+        }
+
+    ///
+    /// Encodes the chars in `src` using `encoder`.
+    ///
+    /// Returns an array of the encoded bytes whose size is the number of bytes encoded.
+    ///
+    def encode(rc: Region[rc], src: Array[Char, r], encoder: CharsetEncoder): Array[Int8, rc] \ IO + rc + r =
+        let numChars = Array.length(src);
+        let dest: Array[Int8, rc] = Array.empty(rc, numChars * 4);
+        let charBuffer = CharBuffer.wrap(src);
+        let byteBuffer = ByteBuffer.wrap(dest);
+        checked_ecast(encoder |> encodeBuffered(rc, charBuffer, byteBuffer))
+
+    ///
+    /// Encodes the chars from `charBuffer` into `byteBuffer` using `encoder`.
+    ///
+    /// Returns an array of the encoded bytes whose size is the number of bytes encoded.
+    ///
+    def encodeBuffered(rc: Region[rc], charBuffer: CharBuffer, byteBuffer: ByteBuffer, encoder: CharsetEncoder): Array[Int8, rc] \ IO + rc =
+        let result = encoder.encode(charBuffer, byteBuffer, false);
+        let bytesEncoded = byteBuffer.position();
+        let dest = byteBuffer.array() |> Array.take(rc, bytesEncoded);
+        if (result.isOverflow()) {
+            charBuffer.flip();
+            byteBuffer.clear();
+            let rest = encoder |> encodeBuffered(rc, charBuffer, byteBuffer);
+            Array.append(rc, dest, rest)
```

Comment:
This branch is only possible if the encoding can have more than 4 bytes per char which is super edge case.

I figured it's good to have this branch just for correctness.

But there's a couple of points to be raised such as:
* i will need to write a custom implementation of a decoder that supports 5 byte chars just to test this (probably doable, essentially just a mock)
* this array.append is not efficient, but introducing a mutable map would have performance cost for regular encodings, which i don't think is worth while

Reply:
seems that possibly it's due to the fact that `averageBytesPerChar` does return a float and they wanted api consistency like if someone is using both values in an equation together. tho i found a JDK ticket talking about fractional values here. I think maybe the best bet is allocate: `numChars * maxBytesPerChar |> ceil |> clampToInt32(nanValue = 4)` (4 is the greatest `maxBytesPerChar` in the java `StandardCharsets`).

It seems like an unrealistic edge case that there would be a charset that breaks this, and then the blame would probably be on the charset and not on us. And, this would be good because we would not need to allocate `4 * numChars` for a charset with, for example, `maxBytesPerChar` of 2.

Thoughts?

Reply:
Seems reasonable.